### PR TITLE
feat: support lists in 'resolve_keyed_by'

### PR DESCRIPTION
**Pull request by ahal**
I tested this in Gecko and it resulted in an identical full graph

### CODE REVIEW

**Review by jcristau — CHANGES_REQUESTED** [review]
Looks like this breaks on python 3.8?

File: src/taskgraph/util/keyed_by.py [hunk 1]
```diff
@@ -2,8 +2,40 @@
...
 
+from typing import Any, Dict, Generator, Tuple
 
-from .attributes import keymatch
+from taskgraph.util.attributes import keymatch
+
+
+def iter_dot_path(
+    container: Dict[str, Any], subfield: str
+) -> Generator[Tuple[dict[str, Any], str], None, None]:
```
Comment (hneiva) [inline src/taskgraph/util/keyed_by.py:12]: Python < 3.9 can't handle type annotation using built-in types 😢 Just need to replace `dict` with `Dict` (I think it can handle `str` tho). Very annoying.

**Comment by bhearsum** [comment]
Moving this back to draft to get it out of review queues.

**Comment by ahal** [comment]
Oops, I forgot about this PR.. I'll get it into shape soon

File: test/test_util_keyed_by.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        pytest.param(
+            {"a": {"b": {"c": 1}}, "d": 2}, "a.b.c", [({"c": 1}, "c")], id="simple case"
+        ),
+        pytest.param(
+            {"a": [{"b": 1}, {"b": 2}, {"b": 3}], "d": 2},
+            "a.b",
+            [({"b": 1}, "b"), ({"b": 2}, "b"), ({"b": 3}, "b")],
+            id="list case",
+        ),
+    ),
```
Comment (bhearsum) [inline test/test_util_keyed_by.py:24]: Some tests for error / corner cases would be welcome. Eg:
container: `{"a": [{"b": 1}, {"c": 2}]}`
subfield: `a.b`

File: docs/howto/resolve-keyed-by.rst [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                      default: bar.tar.gz
+
+With the associated transform:
+
+.. code-block:: python
+
+   @transforms.add
+   def resolve_artifact_paths(config, tasks):
+       for task in tasks:
+           resolve_keyed_by(task, "worker.artifacts.path", task["label"])
```
Comment (bhearsum) [inline docs/howto/resolve-keyed-by.rst:191]: I don't want to bikeshed on this too much, so this is certainly not a blocker, but I have to admit that I find it a bit unintuitive that `.path` accesses multiple list items (in addition to access plain dict keys). I wonder if jq-style syntax would be preferable here? Eg: `worker.artifacts[].path`?

**Review by bhearsum — APPROVED** [review]
The doc improvements here are fantastic!

File: docs/howto/resolve-keyed-by.rst [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+           for platform in task.pop("platforms"):
+               new_task = deepcopy(task)
+               # ...
+               resolve_keyed_by(
+                   new_task,
+                   "worker.max-runtime",
+                   task["label"],
+                   platform=platform,
+               )
+               yield task
```
Comment (jcristau) [inline docs/howto/resolve-keyed-by.rst:150]:
```suggestion
               yield new_task
```

File: docs/howto/resolve-keyed-by.rst (hunk 3 above)
Comment (ahal) [reply docs/howto/resolve-keyed-by.rst:191]: I agree with you. I'll mark this as draft until I get a chance to implement a `[]` syntax.

**Comment by ahal** [comment]
I didn't end up needing this for the thing I was working on at the time, so I've neglected it again.. but I'll try and get this landed at some point.

File: docs/howto/resolve-keyed-by.rst [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+If there are more than one fields that another field may be keyed by, it
+can be specified like this:
+
+.. code-block:: python
+
+   Optional("max-run-time"): optionally_keyed_by("platform", "build-type", int)
+
+
+In this example either ``by-platform`` or ``by-build-type`` may be used. You
+may specify fields as you like this way, as long as the last argument to
```
Comment (ahal) [inline docs/howto/resolve-keyed-by.rst:249]: s/specify fields/specify as many fields

File: docs/howto/resolve-keyed-by.rst (hunk 3 above)
Comment (ahal) [reply docs/howto/resolve-keyed-by.rst:191]: `worker.artifacts[].path` I presume?

**Review by bhearsum — APPROVED** [review]
